### ISSUE TITLE
AttributeError: module 'byteps.torch' has no attribute 'push_pull'

### ISSUE THREAD

**Issue by boscotsang** [bytedance/byteps#35]
I used hps.allreduce and an error was raised
`AttributeError: module 'byteps.torch' has no attribute 'allreduce'`
However, I replace hvd.allreduce with bps.push_pull, there was alos an error
`AttributeError: module 'byteps.torch' has no attribute 'push_pull'`

**Comment by ymjiang** [comment]
For pytorch please use `byteps_push_pull`.

**Comment by boscotsang** [comment]
When I use bps.byteps_push_pull to compute the average loss across multiple process. 
```
def metric_average(val, name):
    tensor = torch.tensor(val)
    avg_tensor = bps.byteps_push_pull(tensor, name=name)
    return avg_tensor.item()

loss_val = train()
loss_val = metric_average(loss_val, name="loss_val")
```
It raised int object has no attribute item()
And after I remove .item(), the following error was raised
`[2019-07-02 17:37:11.589984: F byteps/common/nccl_manager.cc:35] Check failed: e == cudaSuccess || e == cudaErrorCudartUnloadin`
After remove the bps.byteps_push_pull, the error disapppeared.

**Comment by ymjiang** [comment]
@boscotsang Could you provide the complete scripts so that we can try to reproduce?

**Comment by bobzhuyb** [comment]
Using byteps_push_pull is not intended. The problem was that we didn't expose the correct push_pull API in  torch/__init__.py

@boscotsang I have a PR that addresses this (https://github.com/bytedance/byteps/pull/36), We will do our test. Would you also try it in your environment and let us know whether it works? You can use the code in branch "torch_dev".

**Comment by boscotsang** [comment]
@ymjiang  example/pytorch/train_mnist_byteps.py could be a reproduction code since its function metric average used allreduce. 
@bobzhuyb I fetch the torch_dev branch and modify the allreduce with push_all in example/pytorch/train_mnist_byteps.py. Now I can call the push_pull function. However, the error
`[2019-07-03 12:41:43.437152: F byteps/common/nccl_manager.cc:35] Check failed: e == cudaSuccess || e == cudaErrorCudartUnloading CUDA: an illegal memory access was encountered` 
is still exist.

**Comment by bobzhuyb** [comment]
@boscotsang Okay, for this kind of problem, we need to know your cuda version, nccl version, gcc version, and how you installed byteps (from source code, pip package, or docker)

**Comment by boscotsang** [comment]
@bobzhuyb 
CUDA Version: 9.0
GCC Version: 4.9.3
NCCL Version: 2.4.8
I installed byteps from source code, and running single-machine multi-gpu.

**Comment by bobzhuyb** [comment]
Which GPU are you using?

**Comment by boscotsang** [comment]
@bobzhuyb 
NVIDIA P40

**Comment by boscotsang** [comment]
By the way, I have another machine with NVIDIA 1080ti and nccl 2.2.13 and I have also encountered the problem in this machine.

**Comment by bobzhuyb** [comment]
We can't reproduce the `CUDA: an illegal memory access was encountered` error.. We tried gcc 4.9, 5.4, nccl 2.4.8, 2.3.7, pytorch 1.0.1, 1.1.0. Still can't reproduce the error.

Would you paste your detailed commands, and how many GPUs do you have?

**Comment by boscotsang** [comment]
I have 8 gpus and the following is the commands

```
export NVIDIA_VISIBLE_DEVICES=4,5,6,7
export CUDA_VISIBLE_DEVICES=4,5,6,7 
export DMLC_WORKER_ID=0
export DMLC_NUM_WORKER=1
export DMLC_ROLE=worker

export DMLC_NUM_SERVER=1
export DMLC_PS_ROOT_URI=127.0.0.1
export DMLC_PS_ROOT_PORT=1234

python byteps/launcher/launch.py python example/pytorch/train_mnist_byteps.py
```
I commented the train() function in train_mnist_byteps.py in order to directly run the test() function which used metric_average

**Comment by bobzhuyb** [comment]
Ahhh... This could be due to we that have an assumption somewhere that the GPU list starts with 0,1,2... In fact, I already realized a few places we wrote `cudaSetDevice(BytePSGlobal::GetLocalRank())`

Would you try using NVIDIA_VISIBLE_DEVICES=0,1,2,3 ? 

We will fix this.

**Comment by boscotsang** [comment]
I have try NVIDIA_VISIBLE_DEVICES=0,1,2,3 and the problem is still exist

**Comment by boscotsang** [comment]
I guess that in function metric_average, the tensor was constructed in cpu so nccl cannot handle the tensor correctly. I changed the `torch.tensor(val)` to `torch.tensor(val, device=torch.device("cuda")` and the error was gone.

**Comment by boscotsang** [comment]
By the way, the return value of push_pull is the sum of the tensors even if average=True

**Comment by bobzhuyb** [comment]
@boscotsang  Thanks a lot for the report! We will fix the averaging problem.

For CPU tensors, BytePS registers it with NCCL and maps it into GPU memory space. For GPUs that support this, NCCL can still handle it (though it would be slower than real GPU memory). We need to check whether P40 supports this.. We are using T100 and it seems to be fine.

**Comment by boscotsang** [comment]
@bobzhuyb Does the average problem of push_pull influence  the gradient average of DsitributedOptimizer?

**Comment by bobzhuyb** [comment]
@boscotsang 
The average is done here https://github.com/bytedance/byteps/blob/master/byteps/torch/ops.cc#L89

We have tested the convergence of torch examples, and it seemed correct. We need to check why invoking push_pull directly does not average the final result.

**Comment by ymjiang** [comment]
Fixed 0266f392494a2a7e92105dbd3b676e9a2422eb65. 
@boscotsang Thank you very much for the reports.

**Comment by boscotsang** [comment]
It works, thanks
